### DISCUSSION TITLE
timeFilter macro with Azure DataSource on legacy ApplicationInsights datasource references incorrect DateTime field

### DISCUSSION THREAD

**Discussion by ehuggz1** [grafana/grafana#43469] [Data Sources]
**What happened**:
1. Create Panel with Azure DataSource
2.  Select Logs
3.  Select Application Insights resource as the data source
4.  Set Query to past 24 hours
5.  Enter $__timeFilter() macro as part of query - no rows returned
6.  Manually set datetime with "| where timestamp > ago(24h)" rows returned.

Looked at generated query and note that $__timeFilter() is expanding to:

"| where ['TimeGenerated'] >= datetime('2021-12-16T15:54:10Z') and ['TimeGenerated'] <= datetime('2021-12-16T21:54:10Z')"

Note the use of the "TimeGenerated" field.

The Application Insights field for DateTime is "timestamp"

**What you expected to happen**:
$__timeFilter() macro should use "timestamp" as the DateTime field when selecting ApplicationInsights as the data source

Note: Microsoft has indicated the Application Insights data source will be deprecated in 2024. Microsoft provides an option to migrate legacy ApplicationInsights to use a Log Analytics Workspace as the datasource - which would then use the TimeGenerated field as the DateTime field.  I have also seen Microsoft Functionality fall into the same trap, i.e. expecting (or requiring) Kusto queries to use the TimeGenerated field as the DateTime field. 

As an aside:
We've put off migrating AI backend to Log Analytics Workspace as:

a.  The migration orphans old data
b. All of our existing Application Insight queries in Grafana have to be updated to a LogAnalytics Workspace data source.

Need to know if Microsoft has painted everyone into a corner and we just need to bite the bullet and migrate our AI logs to LogAnalytic workspaces.

**How to reproduce it (as minimally and precisely as possible)**:
1. Select Azure Data Source
2. Select (legacy) Application Insights resource (an AI resource that DOES NOT use Log Analytics Workspace as backend)
3.  Create query: "traces | $__timeFilter()
4.  Execute query, select query inspector, select query tab, and view generated query (see screenshot)
**Anything else we need to know?**:

**Environment**:
- Grafana version: Grafana Online (Pro) 8.3.3
- Data source type & version: Azure
- User OS & Browser: Windows 11, Edge
![timeFilter query 2021-12-16_17-05-29](https://user-images.githubusercontent.com/46914110/146456379-371a2dff-289a-4453-a891-0c71b3a32904.jpg)

**Comment by ehuggz** [comment]
@Elfo404 I migrated my Production Application Insights to send new data to a LogAnalytics backend and this is what I found regarding timestamp vs TimeGenerated and the Grafana $__timeFilter() macro expansion:

1.  Microsoft translates the old schema, including "timestamp", into the new Log Analytics Schema when using an Application Insights data source for queries.
    a. As a result, old grafana queries which still expand to timestamp, continue to function and return data.
2.  New grafana queries against the migrated Application Insights resource which expand $__timeFilter() to use "TimeGenerated" as the DateTime stamp, generate errors and return no values.
3. New grafana queries against the new, underlying Log Analytics resource using $__timeFilter() which expand to use TimeGenerated as the DateTime field - function normally and produce data **AFTER** the accompanying query has been updated to conform to the LogAnalytics schema.

In summary:

- Queries against Application Insights using $__timeFilter() fail.
- Queries against the underlying Log Analytics resource for "new" Application Insights resources storing data in Log Analytics Workspaces - work correctly with the Grafana $__timeFilter() macro.

Note: Queries to the underlying LogAnalytics Workspaces have a different schema and only work with $__timeFilter() **AFTER** the accompanying query has been updated to use the new LogAnalytics Schema

Simple Example:

**Successful - Old Grafana Application Insights Query**
`
traces 
| where $__timeFilter()
| where operation_Name == "RouteScheduledTasks" 
| where customDimensions.EventName == "FunctionStarted"
| count
`

**Failure - New Grafana Application Insights Query**
`
traces 
| where $__timeFilter()
| where operation_Name == "RouteScheduledTasks" 
| where customDimensions.EventName == "FunctionStarted"
| count
`

Fails when the $__timeFilter() macro expands to use TimeGenerated as the DateTime field.

**Successful - Log Analytics Query against the Application Insights data now stored in Log Analytics**
`
AppTraces 
| where $__timeFilter()
| where OperationName == "RouteScheduledTasks" 
| where Properties.EventName == "FunctionStarted"
| count
`
Succeeds but requires changes to conform to the LogAnalytics Schema

- traces to AppTraces
- operation_Name to OperationName
- customDimensions to Properties

As it stands after migrating Application Insights to use Log Analytics to store new data: 

- Existing Grafana Application Insights queries using $__timeFilter() continue to work
- New Grafana queries with $__timeFilter() must use Log Analytics as the data source **AND** must use the new LogAnalytics Schema

If Azure Data Source can return to previous functionality, i.e. expanding $__timeFilter to use timestamp as the DateTime field for Application Insights resources - both legacy and new Application Insights can successfully retrieve data using the $__timeFilter() macro.

**Comment by andresmgot** [comment]
Hi @ehuggz,

Thank you for the detailed report! As you mention, the recommended path forward is to use Log Analytics. Since Grafana 8.1, the usage of Application Insights [has been deprecated](https://github.com/grafana/grafana/blob/main/CHANGELOG.md#breaking-changes-5) and it is expected to be completely removed in Grafana 9.0. 

Following that trend, we don't expect to work on new code to cover the legacy Application Insights schema. Having said that, I am going to convert this issue to a feature request, in case we see the demand for it and some contributor wants to contribute.

**Reply by ehuggz** [reply]
@andresmgot The deprecation cited above is the deprecation of the pre-8.1 use of Application Insights as selectable data source - noting these queries must be migrated to the new log data source - which can be used for BOTH Application Insights and Log Analytics queries.  

The Grafana deprecation notice does not specifically say that Application Insights as a data source is deprecated, rather that the Grafana Azure log source and Resource Group/Resource picker must be used to select the data source for Kusto Queries - and then existing queries migrated to this new source. 

We made this change with the migration to 8.1 across all existing Application Insight queries.

I do understand the deprecation, given Microsoft's direction. 
Removing Application Insights as an Kusto query data source option in 9.0 is a reasonable timeframe for change.  

If $__timeFilter(timestamp) works with Application Insights then perhaps a Grafana deprecation notice about the use of Application Insights as a Kusto Query Source in general, and the use of $__timeFilter(timestamp) instead of $__timeFilter() would be helpful to other Azure users.

**Reply by andresmgot** [reply]
I am not sure I follow, as far as I can see you can use ` $__timeFilter(timestamp)` using Log Analytics:

![Screenshot from 2021-12-23 09-41-55](https://user-images.githubusercontent.com/4025665/147213678-4be7026e-bc35-440e-bc23-abb84af1594b.png)


So for this:

> Failure - New Grafana Application Insights Query
traces | where $__timeFilter() | where operation_Name == "RouteScheduledTasks" | where customDimensions.EventName == "FunctionStarted" | count
> Fails when the $__timeFilter() macro expands to use TimeGenerated as the DateTime field.


The query could be:

```
traces | where $__timeFilter(timestamp) | where operation_Name == "RouteScheduledTasks" | where customDimensions.EventName == "FunctionStarted" | count
```

Or wouldn't that work?

**Comment by ehuggz** [comment]
@andresmgot You would use $__timeFilter(timestamp) in ApplicationInsights, the datetime field in LogAnalytics is TimeGenerated.

The reason for the request is this:
1. Grafana 8.1 required updating all Azure Application Insight and Log Analytics queries to use the new log data source. Which was a considerable effort.
2. AI data stored in LogAnalytics uses a different schema.
3. Microsoft went to a lot of effort to make the migration easier by auto translating AI query schema to LA after migration.
4. Migrating existing queries to the LogAnalytics schema will be more timeconsuming than the Grafana 8.1 migration.
5.  Grafana and Grafana Customers needed to do nothing to support the Application Insights deprecation until the change to default $__timeFilter() to TimeGenerated.

But - you are correct. $__timeFilter(timestamp) will work with Application Insights queries. But I did not find the functionality well documented and was not aware that was an option until the suggestion to use it. As a result, I am suggesting the next release notes include a breaking change be noted for $__timeFilter() when used with Application Inisights to indicate $__timeFilter(timestamp) is now required, and a note that Application Insights based queries will likely be no longer be supported in 9.0 - which is (hopefully) a shorter time frame than Microsoft's 2024 timeframe. Hopefully because of bigger and better things in 9.0 :-)

**Reply by andresmgot** [reply]
Got it. I don't think the release notes are the right place though: We should not edit old release notes (because it's an automated process) and it wouldn't make sense to add that in the next release notes since there are no changes related to this version. We can add that as a note in the normal documentation. Maybe this document is a good place: https://grafana.com/docs/grafana/latest/datasources/azuremonitor/deprecated-application-insights/

Would you mind opening a PR with what you think would be a good addition? Thanks!

**Comment by ehuggz** [comment]
I'll give that a shot.

**Comment by andresmgot** [comment]
I believe this is no longer relevant with Grafana 9.0. Feel free to open a new discussion if this is still relevant!
